### PR TITLE
Implement AnimState

### PR DESCRIPTION
**Pull request by jonschz**
All the new functions match to 100 %.

I'd appreciate some feedback on the naming:
* `AnimState::m_unk0x10` clearly stores a dynamically sized array, and the values have the same semantics as `AnimInfo::m_unk0x22`. Would it make sense to rename this array, and what would a good name be?
* `AnimState::m_unk0x0c` clearly stores the length of `AnimState::m_unk0x10`.
* `AnimInfo::m_locationsFlags` is subject to the same issue - the array structure is clear, but the semantics of the stored data are not. I added a name here which should be changed in the future.
* If you have better name suggestions for `AnimState::InitFromAnims` and `AnimState::CopyToAnims`, I'd appreciate to hear them.

### CODE REVIEW

**Comment by BJNFNE** [comment]
Looks good to me

File: LEGO1/lego/legoomni/include/legonavcontroller.h [hunk 1]
```diff
@@ -74,6 +74,7 @@ class LegoNavController : public MxCore {
 	);
 	static MxResult UpdateLocation(MxU32 p_location);
 	static MxResult UpdateLocation(const char* p_location);
+	static MxS32 GetNumLocations(void);
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/include/legonavcontroller.h:77]: Should be `GetNumLocations()`

File: LEGO1/lego/legoomni/src/common/legoanimationmanager.cpp [hunk 2]
```diff
@@ -2828,26 +2828,59 @@ AnimState::AnimState()
...
-	// TODO
+	if (m_unk0x10 == NULL) {
+		m_unk0x0c = p_animsLength;
+		m_unk0x10 = new MxU16[p_animsLength];
+		MxS32 numLocations = LegoNavController::GetNumLocations();
+		m_locationsFlagsLength = numLocations;
+		m_locationsFlags = new MxBool[numLocations];
+	}
+
+	this->m_extraCharacterId = p_extraCharacterId;
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/common/legoanimationmanager.cpp:2872]: shouldn't use `this->`

File: LEGO1/lego/legoomni/src/common/legoanimationmanager.cpp [hunk 3]
```diff
@@ -2828,26 +2828,59 @@ AnimState::AnimState()
...
+
+	for (MxS32 i = 0; i < m_unk0x0c; i++) {
+		m_unk0x10[i] = p_anims[i].m_unk0x22;
+	}
+
+	for (MxS32 j = 0; j < m_locationsFlagsLength; j++) {
+		LegoLocation* location = LegoNavController::GetLocation(j);
+		if (location != NULL) {
+			this->m_locationsFlags[j] = location->m_unk0x5c;
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/common/legoanimationmanager.cpp:2881]: shouldn't use `this->`

File: LEGO1/lego/legoomni/include/legoanimationmanager.h [hunk 4]
```diff
@@ -71,22 +71,23 @@ class AnimState : public LegoState {
 	MxBool SetFlag() override;                     // vtable+0x18
 	MxResult Serialize(LegoFile* p_file) override; // vtable+0x1c
 
-	void FUN_100651d0(MxU32, AnimInfo*, MxU32&);
-	void FUN_10065240(MxU32, AnimInfo*, MxU32);
+	void CopyToAnims(MxU32 p_unused, AnimInfo* p_anims, MxU32& p_outExtraCharacterId);
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/include/legoanimationmanager.h:74]: if it's unused, don't give any name at all - just leave it `MxU32`

File: LEGO1/lego/legoomni/src/common/legoanimationmanager.cpp [hunk 5]
```diff
@@ -2828,26 +2828,59 @@ AnimState::AnimState()
...
 {
-	// TODO
+	delete[] m_unk0x10;
+	delete[] m_locationsFlags;
 }
 
-// STUB: LEGO1 0x100651d0
-void AnimState::FUN_100651d0(MxU32, AnimInfo*, MxU32&)
+// FUNCTION: LEGO1 0x100651d0
+void AnimState::CopyToAnims(MxU32 p_unused, AnimInfo* p_anims, MxU32& p_outExtraCharacterId)
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/common/legoanimationmanager.cpp:2843]: if it's unused, don't give any name at all - just leave it `MxU32`

File: LEGO1/lego/legoomni/src/common/legoanimationmanager.cpp [hunk 6]
```diff
@@ -2864,57 +2897,76 @@ MxResult AnimState::Serialize(LegoFile* p_file)
...
+		}
+
+		for (MxS32 j = 0; j < m_locationsFlagsLength; j++) {
+			if (LegoNavController::GetLocation(j) != NULL) {
+				m_locationsFlags[j] = 0;
+			}
+		}
+
+		return TRUE;
+	}
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/common/legoanimationmanager.cpp:2970]: add a new line before the last return

File: LEGO1/lego/legoomni/src/entity/legonavcontroller.cpp [hunk 7]
```diff
@@ -482,6 +482,12 @@ LegoLocation* LegoNavController::GetLocation(MxU32 p_location)
 	return NULL;
 }
 
+// FUNCTION: LEGO1 0x10055740
+MxS32 LegoNavController::GetNumLocations(void)
```
Comment (foxtacles) [inline LEGO1/lego/legoomni/src/entity/legonavcontroller.cpp:486]: Should be `GetNumLocations()`

**Review by foxtacles — COMMENTED** [review]
Naming looks alright to me for now. To establish proper names we need to understand the actual purpose of this code in the game

**Comment by jonschz** [comment]
Thanks for the review! For what it's worth, it appears that this class was changed between Beta 1.0 and release. As far as I can tell, the only interesting function in Beta 1.0's AnimState's vtable is `AnimState::Serialize`.

**Pull request merged**